### PR TITLE
Fixed #20223 -- Added keep_lazy() as a replacement for allow_lazy().

### PR DESCRIPTION
**Pull request by uruz**
The semantics stay the same but the new syntax allow the use of
keep_lazy as a decorator with the @-syntax.
It also makes sure that type parameters are passed to lazy,
which would result in an error otherwise.

The cleanups introduced in this change also incidentally fix #20222.

### CODE REVIEW

**Comment by uruz** [comment]
This is an updated version of @bmispelon's #1007

File: django/utils/functional.py [hunk 1]
```diff
@@ -174,23 +176,44 @@ def _lazy_proxy_unpickle(func, args, kwargs, *resultclasses):
 
 
 def allow_lazy(func, *resultclasses):
+    warnings.warn(
+        "django.utils.functional.allow_lazy will be removed in Django 1.8. "
```
Comment (coldmind) [inline django/utils/functional.py]: `removed in Django 2.1`, if `RemovedInDjango21Warning`

File: docs/ref/utils.txt [hunk 2]
```diff
@@ -514,31 +514,54 @@ Atom1Feed
...
+    just pass the ``unicode`` type to ``keep_lazy`` (or even simpler, use the
+    ``keep_lazy_text`` decorator described in the next section).
 
     Using this decorator means you can write your function and assume that the
     input is a proper string, then add support for lazy translation objects at
     the end.
 
+.. function:: keep_lazy_text(func)
+
+    A shortcut for ``keep_lazy(unicode)``.
```
Comment (charettes) [inline docs/ref/utils.txt]: Not sure if we should refer to it as `str` (implying Python3) or `six.text_type`, `unicode` looks confusing to me.

File: django/utils/functional.py (hunk 1 above)
Comment (uruz) [inline django/utils/functional.py]: Missed that occurience of "1.8", thanks. Fixed by force-pushing

File: docs/ref/utils.txt (hunk 2 above)
Comment (uruz) [inline docs/ref/utils.txt]: I've replaced unicode with six.text_type, thanks

File: docs/ref/utils.txt [hunk 3]
```diff
@@ -501,7 +501,7 @@ Atom1Feed
         z = person.friends         # does not call
         x is z                     # is True
 
-.. function:: allow_lazy(func, *resultclasses)
```
Comment (timgraham) [inline docs/ref/utils.txt]: Please keep at least a stub of documentation for `allow_lazy()` and mark it deprecated.

File: docs/ref/utils.txt [hunk 4]
```diff
@@ -514,31 +514,55 @@ Atom1Feed
...
+    A common use-case is to have functions that return text. For these, you can
+    just pass the ``six.text_type`` type to ``keep_lazy`` (or even simpler, use the
+    ``keep_lazy_text`` decorator described in the next section).
 
     Using this decorator means you can write your function and assume that the
     input is a proper string, then add support for lazy translation objects at
     the end.
 
+.. function:: keep_lazy_text(func)
+
```
Comment (timgraham) [inline docs/ref/utils.txt]: missing `.. versionadded:: 1.9`

**Comment by timgraham** [comment]
Please see https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature for everything needed for the deprecation.

File: tests/template_tests/filter_tests/test_escapejs.py [hunk 5]
```diff
@@ -51,3 +53,9 @@ def test_paragraph_separator(self):
             escapejs_filter('paragraph separator:\u2029and line separator:\u2028'),
             'paragraph separator:\\u2029and line separator:\\u2028',
         )
+
+    def test_lazy_string(self):
+        append_script = lazy(lambda string: r'<script>this</script>' + string, six.text_type)
+        self.assertEqual(escapejs_filter(append_script('whitespace: \r\n\t\v\f\b')),
```
Comment (timgraham) [inline tests/template_tests/filter_tests/test_escapejs.py]: Please use hanging indent style as above test.

File: docs/topics/performance.txt [hunk 6]
```diff
@@ -223,7 +223,7 @@ QuerySet <when-querysets-are-evaluated>`. Avoiding the premature evaluation of
 a ``QuerySet`` can save making an expensive and unnecessary trip to the
 database.
 
-Django also offers an :meth:`~django.utils.functional.allow_lazy` decorator.
+Django also offers an :meth:`~django.utils.functional.keep_lazy` decorator.
```
Comment (timgraham) [inline docs/topics/performance.txt]: an -> a

File: docs/releases/1.9.txt [hunk 7]
```diff
@@ -268,6 +268,11 @@ Miscellaneous
   :class:`~django.forms.SelectDateWidget` in ``django.forms.widgets``
   (or simply ``django.forms``) instead.
 
+* ``django.utils.functional.allow_lazy`` is deprecated.
+  This function has been deprecated in favor of a new :func:`keep_lazy<django.utils.functional.keep_lazy>`
+  one which can be used with a more natural decorator syntax.
+
```
Comment (timgraham) [inline docs/releases/1.9.txt]: extra newline

File: docs/releases/1.9.txt [hunk 8]
```diff
@@ -268,6 +268,11 @@ Miscellaneous
   :class:`~django.forms.SelectDateWidget` in ``django.forms.widgets``
   (or simply ``django.forms``) instead.
 
+* ``django.utils.functional.allow_lazy`` is deprecated.
+  This function has been deprecated in favor of a new :func:`keep_lazy<django.utils.functional.keep_lazy>`
```
Comment (timgraham) [inline docs/releases/1.9.txt]: please wrap at 79 chars, you can have a linkebreak after :func:`keep_lazy

File: tests/decorators/tests.py [hunk 9]
```diff
@@ -8,7 +8,7 @@
 from django.http import HttpRequest, HttpResponse, HttpResponseNotAllowed
 from django.middleware.clickjacking import XFrameOptionsMiddleware
 from django.utils.decorators import method_decorator
-from django.utils.functional import allow_lazy, lazy
```
Comment (timgraham) [inline tests/decorators/tests.py]: we should probably keep at least once test for `allow_lazy` to ensure we don't break it during the deprecation period.

**Comment by timgraham** [comment]
@uruz, do you have time/interest in completing this soon? Was thinking we could offer it as a task for the DjangoconEU sprints if not.

**Comment by uruz** [comment]
@timgraham I know it's my debt, but I will have time to work on it not earlier than two weeks from now. I suppose DjangoconEU sprints will happen earlier, so I don't mind if this issue would be offered and hopefully resolved during Djangocon sprints.

**Comment by timgraham** [comment]
Closing due to inactivity. Please send a new PR if you find time to return to it, thanks!

**Pull request closed without merging**